loader/include/Geode/utils/Task.hpp

```diff
@@ -825,39 +888,42 @@ namespace geode {
-                                        }
-                                        else if (event->isCancelled()) {
-                                            NewTask::cancel(handle.lock());
-                                        }
+                            handle.lock()->m_extraData =
+                                std::make_unique<typename NewTask::Handle::ExtraData>(
+                                    // make the second event listener that waits for the mapper's
+                                    // task and just forwards everything through
+                                    static_cast<void*>(new EventListener<NewTask>(
+                                        [handle](typename NewTask::Event* event) mutable {
+                                            if (auto v = event->getValue()) {
+                                                NewTask::finish(handle.lock(), std::move(*v));
+                                            }
+                                            else if (auto p = event->getProgress()) {
+                                                NewTask::progress(handle.lock(), std::move(*p));
+                                            }
+                                            else if (event->isCancelled()) {
+                                                NewTask::cancel(handle.lock());
+                                            }
+                                        },
+                                        std::move(newInnerTask)
+                                    )),
+                                    +[](void* ptr) {
+                                        delete static_cast<EventListener<NewTask>*>(ptr);
                                     },
-                                    std::move(newInnerTask)
-                                )),
-                                +[](void* ptr) {
-                                    delete static_cast<EventListener<NewTask>*>(ptr);
-                                },
-                                +[](void* ptr) {
-                                    static_cast<EventListener<NewTask>*>(ptr)->getFilter().cancel();
-                                }
-                            );
+                                    +[](void* ptr) {
+                                        static_cast<EventListener<NewTask>*>(ptr)->getFilter().cancel();
+                                    }
+                                );
                         }
                         else if (auto p = event->getProgress()) {
-                            // no guarantee P and NewProgress are compatible
-                            // nor does it seem like the intended behavior?
-                            // TODO: maybe add a mapper for progress?
+                            // Map progress from P to NewProgress if possible
+                            if constexpr (std::is_convertible_v<P, NewProgress>) {

```

Comment:
This can lead to weird behaviour.

Let's say we're chaining 2 web requests.

If I understand this correctly, then progress will roll up to 100, then back to 0 and up to 100 again. Not good.

Reply:
how could i fix that?

Reply:
plan here was to also create a mapper for progress, though there's no easy solution here. most chaining implementations don't care about the progress of the tasks before the final one.

Reply:
lemme try something

Reply:
@Fleeym would something like this work?

```cpp
if constexpr (std::is_convertible_v<P, NewProgress>) {
    auto scaledProgress = static_cast<float>(*p) * 0.5f;
    NewTask::progress(handle.lock(), NewProgress(scaledProgress));
}
```

Reply:
the progress doesnt have to be a number, but also you can chain more than 2 tasks

Reply:
@Fleeym what about something like this? probably not since i literally just came up with it but i might be able to get something from it

```cpp
             if constexpr (std::is_convertible_v<P, NewProgress>) {
                                size_t totalTasks = 1;
                                auto currentHandle = handle.lock();
                                while (currentHandle && currentHandle->m_extraData) {
                                    totalTasks++;
                                    auto listener = static_cast<EventListener<NewTask>*>(currentHandle->m_extraData->m_data);
                                    currentHandle = listener->getFilter().m_handle;
                                }
                                
                                auto scaledProgress = static_cast<float>(*p) / static_cast<float>(totalTasks);
                                NewTask::progress(handle.lock(), NewProgress(scaledProgress));
                            }
                    }
```

Reply:
find a way to count how many tasks are happening at once, then scale the progress tracker to that amount so that it doesn't restart the tracker every time a task ends

Reply:
well progress isn't guaranteed to be a float, so this can't really work.

progress can be anything from an enum to a primitive type to a full blown class

Reply:
oh
ok yeah i see what you mean by there's no easy solution

Reply:
gimme about 2 hours and i'll see if i can come up with something, otherwise i'll just revert it to its original state

Reply:
assuming you want to linearly interpolate between enum values is crazy 😭 this feature seems overkill

Reply:
> assuming you want to linearly interpolate between enum values is crazy 😭 this feature seems overkill

i mean this pr is meant to implement all todos and this was something that was in a todo so

Reply:
maybe some todos are not worth it, like explained in the comment itself and flame's msg

Reply:
but i don't see the problem in implementing this, especially considering the fact that my proposed solution works with (as Fleeym said) primitives, enums, and classes?

Reply:
is it really worth it when the behavior can be so arbitrary? (and also doing the task chain count is hacky + expensive)

Reply:
> is it really worth it when the behavior can be so arbitrary? (and also doing the task chain count is hacky + expensive)

one of the problems fleeym had with my original implementation was that the progress would reset for each task, the task count is just so that i can scale the progress tracker

Reply:
In my opinion we should leave this as is, as changing this means an API break. Also the more obvious functionality here is just tracking the progress of the last task.

We'll keep in mind the implementation for a future major version though.

Reply:
oki